packages/tangle/tipmanager.go

```diff
@@ -324,9 +327,207 @@ func (t *TipManager) Tips(p payload.Payload, countParents int) (parents MessageI
+
+	t.tangle.Booker.MarkersManager.Manager.Sequence(marker.SequenceID()).Consume(func(sequence *markers.Sequence) {
+		// If there is a confirmed marker before the oldest unconfirmed marker, and it's older than minSupportedTimestamp, need to walk message past cone of oldestUnconfirmedMarker.
+		if sequence.LowestIndex() < oldestUnconfirmedMarker.Index() {
+			confirmedMarkerIdx := t.getPreviousConfirmedIndex(sequence, oldestUnconfirmedMarker.Index()-1)
+
+			confirmedMarkerMsg := t.getMarkerMessage(markers.NewMarker(sequence.ID(), confirmedMarkerIdx))
+			if t.tangle.ConfirmationOracle.IsMessageConfirmed(confirmedMarkerMsg.ID()) && confirmedMarkerMsg.IssuingTime().Before(minSupportedTimestamp) {
+				messageWalker.Push(t.tangle.Booker.MarkersManager.MessageID(oldestUnconfirmedMarker))
+			}
+		}
+
+		// process markers from different sequences that are referenced by current marker's sequence, i.e., walk the sequence DAG
+		referencedMarkers := sequence.ReferencedMarkers(marker.Index())
+		referencedMarkers.ForEach(func(sequenceID markers.SequenceID, index markers.Index) bool {
+			referencedMarker := markers.NewMarker(sequenceID, index)
+			// if referenced marker is confirmed and older than minSupportedTimestamp, walk unconfirmed message past cone of oldestUnconfirmedMarker
+			referencedMarkerMsg := t.getMarkerMessage(markers.NewMarker(sequenceID, index))
+			if t.tangle.ConfirmationOracle.IsMessageConfirmed(referencedMarkerMsg.ID()) && referencedMarkerMsg.IssuingTime().Before(minSupportedTimestamp) {
+				messageWalker.Push(t.tangle.Booker.MarkersManager.MessageID(oldestUnconfirmedMarker))
+				return true
+			}
+			// otherwise, process the referenced marker
+			markerWalker.Push(*referencedMarker)
+			return true
+		})
+	})
+	return true
+}
+
+func (t *TipManager) processMarker(pastMarker *markers.Marker, minSupportedTimestamp time.Time, oldestUnconfirmedMarker *markers.Marker) (tscValid bool) {
+	// oldest unconfirmed marker is in the future cone of the past marker (same sequence), therefore past marker is confirmed and there is no need to check
+	if pastMarker.Index() < oldestUnconfirmedMarker.Index() {
+		return true
+	}
+
+	return t.getMarkerMessage(oldestUnconfirmedMarker).IssuingTime().After(minSupportedTimestamp)
+}
+
+func (t *TipManager) checkMessage(messageID MessageID, messageWalker *walker.Walker, minSupportedTimestamp time.Time) (timestampValid bool) {
+	timestampValid = true
+
+	if t.tangle.ConfirmationOracle.IsMessageConfirmed(messageID) {
+		return
+	}
+	t.tangle.Storage.Message(messageID).Consume(func(message *Message) {
+		if message.IssuingTime().Before(minSupportedTimestamp) {
+			timestampValid = false
+			messageWalker.StopWalk()
+			return
+		}
+
+		// walk through strong parents' past cones
+		for parentID := range message.ParentsByType(StrongParentType) {
+			messageWalker.Push(parentID)
+		}
+	})
+	return timestampValid
+}
+
+func (t *TipManager) getMarkerMessage(marker *markers.Marker) (markerMessage *Message) {
```

Comment:
This is a bit dangerous and not a good practice imo, as we consume the message and store the pointer to it and return it. However, we tell the object storage to release the cached message object at any time. 

For messages it is not that tragic because we basically just read from it but in general this is a recipe for race conditions. I think it would be better to consume the message and return the properties that we're interested in. Here it should be ID and issuing time.

Reply:
Changed it